File: languages/java/src/main/java/com/bloock/sdk/entity/key/KeyProtectionLevel.java
package com.bloock.sdk.entity.key;

import com.bloock.sdk.bridge.proto.BloockKeysEntities;

/**
 * Represents the protection level of a cryptographic key.
 */
public enum KeyProtectionLevel {
  /**
   * Indicates that the key is protected by software.
   */
  SOFTWARE,
  /**
   * Indicates that the key is protected by a Hardware Security Module (HSM).
   */
  HSM,
  UNRECOGNIZED;

  public static KeyProtectionLevel fromProto(BloockKeysEntities.KeyProtectionLevel protection) {
    switch (protection) {
      case SOFTWARE:
        return KeyProtectionLevel.SOFTWARE;
      case HSM:
        return KeyProtectionLevel.HSM;
      default:
        return KeyProtectionLevel.UNRECOGNIZED;
    }
  }

  public BloockKeysEntities.KeyProtectionLevel toProto() {
    switch (this) {
      case SOFTWARE:
        return BloockKeysEntities.KeyProtectionLevel.SOFTWARE;
      case HSM:
        return BloockKeysEntities.KeyProtectionLevel.HSM;
      default:
        return BloockKeysEntities.KeyProtectionLevel.UNRECOGNIZED;
    }
  }
}

File: languages/js/src/entity/key/key_protection_level.ts
import * as keysEntitiesProto from "../../bridge/proto/keys_entities";

/**
 * Represents the protection level of a cryptographic key.
 */
export enum KeyProtectionLevel {
  /**
   * Indicates that the key is protected by software.
   */
  SOFTWARE,
  /**
   * Indicates that the key is protected by a Hardware Security Module (HSM).
   */
  HSM
}

export namespace KeyProtectionLevel {
  export function toProto(
    type: KeyProtectionLevel
  ): keysEntitiesProto.KeyProtectionLevel {
    switch (type) {
      case KeyProtectionLevel.SOFTWARE:
        return keysEntitiesProto.KeyProtectionLevel.SOFTWARE;
      case KeyProtectionLevel.HSM:
        return keysEntitiesProto.KeyProtectionLevel.HSM;
    }
  }

  export function fromProto(
    type: keysEntitiesProto.KeyProtectionLevel | undefined
  ): KeyProtectionLevel {
    switch (type) {
      case keysEntitiesProto.KeyProtectionLevel.SOFTWARE:
        return KeyProtectionLevel.SOFTWARE;
      case keysEntitiesProto.KeyProtectionLevel.HSM:
        return KeyProtectionLevel.HSM;
      default:
        return KeyProtectionLevel.SOFTWARE;
    }
  }
}
